File: src/node/zoy/simple.mjs
import zmq from 'zeromq'; // note needs npm install zeromq@6.0.0-beta.19 (or higher)
import { inflateSync, unzipSync, gunzipSync } from 'zlib';

var sock;

// VPS Host
// const host = 'tcp://infomancer.uk'
// const port = 5551;

// Local testing
const host = 'tcp://localhost'
//const host = 'tcp://192.168.1.137'
const port = 5555;

const zmqURL = `${host}:${port}`;

async function run() {
  sock = new zmq.Subscriber;

 sock.connect(`${zmqURL}`);

  // Galaxy Tick - one per day usually
  sock.subscribe('GalaxyTick');
  
  // System Tick - every detected system change, ~7k per day
  sock.subscribe('SystemTick');

  // FactionChanges - factions arriving in and/or retreating from a System ~50 per day
  sock.subscribe('FactionChanges');
  
  // FactionExpanded - report of the System a faction expanded from
  sock.subscribe('FactionExpandedFrom');
  
  sock.subscribe('Heartbeat'); // every 5 minutes

  console.log(`Subscriber connected to ${zmqURL}`);

  for await (const [topic, msg] of sock) {
    const sTopic = topic.toString();
    const payload = msg;
    switch (sTopic) {
      // case 'GalaxyTick':
      //   processGalaxytick(payload);
      //   break;
      case 'SystemTick':
        processSystemTick(payload);
        break;
      // case 'FactionChanges':
      //   processFactionChanges(payload);
      //   break;
      // case 'FactionExpandedFrom':
      //   processFactionExpandedFrom(payload);
      //   break;
      case 'Heartbeat':
        processHeartbeat(payload);
        break;
      case 'HeartbeatZ':
        processHeartbeatZ(payload);
        break;
      default:
        console.log(`Unhandled topic: ${sTopic}`);
        break;
    };
  };
};

run();

function processFactionChanges(payload) { console.log(`FactionChanges: ${payload}`) };
function processFactionExpandedFrom(payload) { console.log(`FactionExpandedFrom: ${payload}`) };
function processGalaxytick(payload) { console.log(`Galaxytick: ${payload}`) };
function processSystemTick(payload) { console.log(`SystemTick: ${payload}`) };
function processHeartbeat(payload) { console.log(`Heartbeat: ${payload}`) };
// function processHeartbeatZ(payload) { console.log(`HeartbeatZ: ${inflateSync(payload)}`) };
function processHeartbeatZ(payload) { 
  console.log(`inflate HeartbeatZ: ${inflateSync(payload)}`);
  console.log(`unzip HeartbeatZ: ${unzipSync(payload)}`);
 };


process.on('SIGINT', () => {
  sock.close();
  console.log("Subscriber shutting down...");
});

File: src/python3/zoy/simple.py
import logging

import zmq
import zlib

# Local testing
# HOST = "tcp://localhost"
# PORT = 5555

# live Host
HOST = 'tcp://infomancer.uk'
PORT = 5551;

# Define the topics to subscribe to
SUBSCRIPTIONS = [
    "GalaxyTick",
    "SystemTick",
    "FactionChanges",
    "FactionExpandedFrom",
    "Heartbeat"
]

def run():
    # print(zlib.__version__)
    context = zmq.Context()
    sock = context.socket(zmq.SUB)

    sock.connect(f"{HOST}:{PORT}")

    logging.info(f"Subscriber connected to {HOST}:{PORT}")

    for topic in SUBSCRIPTIONS:
        sock.subscribe(topic.encode())  # Encode topic for byte stream

    while True:
        msg = sock.recv_multipart()  # Receive a single message
        topic, compressed_payload = msg
        topic_str = topic.decode()
        payload = zlib.decompress(compressed_payload).decode()
        logging.info(f"{topic_str}: {payload}")


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    run()

